### PR TITLE
Syntax fixes

### PR DESCRIPTION
**Pull request by Slimakanzer**
- [x] Fix memory leaks when repeatedly opening Radeon Asm documents
- [x] Reduce memory usage
- [x] Asynchronous options initialization on visual studio startup
- [x] Add saving and restoring the current instruction set
- [x] Fix label reference highlighting
- [x] Fix function arguments highlighting
- [x] Fix single-line comments highlighting if there is some token to the left of it
- [x] Fix number highlighting with '-' symbol

### CODE REVIEW

File: VSRAD.Syntax/Core/DocumentFactory.cs [hunk 1]
```diff
@@ -92,71 +86,89 @@ public IDocument GetOrCreateDocument(ITextBuffer buffer)
...
+            if (!_documents.TryGetValue(document.Path, out var documentPair))
+                return;
+
+            documentPair.Key.FileActionOccurred -= TextDocumentActionOccurred;
+            _documents.Remove(document.Path);
+            DocumentDisposed?.Invoke(document);
         }
 
-        private void OnChangeActivatedWindow(Window GotFocus, Window LostFocus)
+        private void OnChangeActivatedWindow(Window gotFocus, Window _)
```
Comment (johnny-keker) [inline VSRAD.Syntax/Core/DocumentFactory.cs:150]: From the context of using, it seems to me that this should be called `OnWindowActivatedChanged`?

File: VSRAD.Syntax/Core/IDocumentAnalysis.cs [hunk 2]
```diff
@@ -10,6 +10,9 @@ public interface IDocumentAnalysis
     {
         IAnalysisResult CurrentResult { get; }
         Task<IAnalysisResult> GetAnalysisResultAsync(ITextSnapshot textSnapshot);
+        void Rescan(RescanReason rescanReason, CancellationToken cancellationToken);
+        void OnDestroy();
+
```
Comment (johnny-keker) [inline VSRAD.Syntax/Core/IDocumentAnalysis.cs:15]: Extra newline

File: VSRAD.Syntax/Guide/IndentGuide.cs [hunk 3]
```diff
@@ -72,32 +77,32 @@ private void AnalysisUpdated(IAnalysisResult analysisResult, RescanReason reason
...
+                UpdateIndentGuides();
+            else
+                CleanupIndentGuidesAsync().ConfigureAwait(false);
         }
 
-        private void CanvasRemoved(object tag, UIElement element) =>
-            _layer.AddAdornment(AdornmentPositioningBehavior.OwnerControlled, null, null, _canvas, CanvasRemoved);
+        private void CanvasRemoved(object tag, UIElement element)
+        {
+            // TODO: here is some code
```
Comment (johnny-keker) [inline VSRAD.Syntax/Guide/IndentGuide.cs:104]: Future work?

File: VSRAD.Syntax/Core/DocumentAnalysis.cs [hunk 4]
```diff
@@ -37,31 +38,50 @@ public async Task<IAnalysisResult> GetAnalysisResultAsync(ITextSnapshot textSnap
...
-                    .Cast<IncludeToken>()
-                    .Select(i => i.Document)
-                    .ToList();
-
-                var analysisResult = new AnalysisResult(parserResult, includes, tokenizerResult.Snapshot);
+                //var includes = parserResult.RootBlock.Tokens
+                //    .Where(t => t.Type == RadAsmTokenType.Include)
+                //    .Cast<IncludeToken>()
+                //    .Select(i => i.Document)
+                //    .ToList();
```
[the remark this answers is not in the document]
Comment (Slimakanzer) [reply VSRAD.Syntax/Core/DocumentAnalysis.cs]: Thank you, fixed in 7c9f046

**Review by johnny-keker — APPROVED** [review]
Done some testing:

* Selected instruction set is restored on reopening VS;
* No `Extension is slowing down VS Startup...` messages;
* Function arguments highlighting seems to work as expected;
* Single line comments without whitespace on the left side highlighted;
* Numbers with `-` delimiters highlighted.

LGTM, works as expected

File: VSRAD.Syntax/Core/DocumentFactory.cs [hunk 5]
```diff
@@ -26,61 +27,54 @@ internal partial class DocumentFactory : IDocumentFactory
...
-            if (buffer.Properties.TryGetProperty(typeof(IDocument), out IDocument document))
-                return document;
-
-            if (!buffer.Properties.TryGetProperty(typeof(ITextDocument), out ITextDocument textDocument))
+            if (!buffer.GetTextDocument(out var textDocument))
                 return null;
+            if (_documents.TryGetValue(textDocument.FilePath, out var documentPair) && textDocument == documentPair.Key)
+                return documentPair.Value;
+
+            // TODO: fix true constant
```
[the remark this answers is not in the document]
Comment (Slimakanzer) [reply VSRAD.Syntax/Core/DocumentFactory.cs:68]: Yes, it's for the feature that is currently not needed

File: VSRAD.Syntax/Core/IDocumentAnalysis.cs (hunk 2 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/Core/IDocumentAnalysis.cs:15]: Thank you, fixed in 7c9f046

File: VSRAD.Syntax/Guide/IndentGuide.cs (hunk 3 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/Guide/IndentGuide.cs]: Forget to implement, fixed in 7c9f046

**Pull request merged**